### PR TITLE
Combination

### PR DESCRIPTION
**Pull request by nirtiac**
Structural overhaul. Dataset needs work.

### CODE REVIEW

**Comment by priyatharsan** [comment]
Merging and then proceeding from here.

File: vulcanai2/models/metrics.py [hunk 1]
```diff
@@ -1,32 +1,66 @@
...
+            self.confusion_matrix += self._fast_hist(
+                prediction.flatten(), target.flatten(), self.num_class)
+
+    def get_scores(self):
+        """Returns accuracy score evaluation result.
+            - overall accuracy
+            - mean accuracy
+            - mean IU
+        """
+        hist = self.confusion_matrix
```
Comment (rfratila) [inline vulcanai2/models/metrics.py]: We are missing a few metrics here like sensitivity, specificity, npc, ppv, etc that can be found in the original Vulcan

File: vulcanai2/datasets/utils.py [hunk 2]
```diff
@@ -107,6 +112,15 @@ def __call__(self, data):
 
 #THIS IS FROM SNEHA  https://github.com/sneha-desai
 def stitch_datasets(df_list, on, index_list=None):
```
Comment (rfratila) [inline vulcanai2/datasets/utils.py]: This will need to be updated with what Joseph uses for dataset stitching.

Comment (rfratila) [reply vulcanai2/datasets/utils.py]: it's too slow and not able to handle larger datasets in it's current state

File: vulcanai2/models/metrics.py [hunk 3]
```diff
@@ -1,32 +1,66 @@
...
+    def _fast_hist(self, prediction, target, num_class):
+        mask = (target >= 0) & (target < num_class)
+        hist = np.bincount(
+                            num_class * target[mask].astype(int) + prediction[mask],
+                            minlength=num_class ** 2,
+                            ).reshape(num_class, num_class)
+        return hist
+    def update(self, predictions, targets):
+        for prediction, target in zip(predictions, targets):
+            self.confusion_matrix += self._fast_hist(
```
Comment (rfratila) [inline vulcanai2/models/metrics.py]: you can just use sklearn to generate the confusion matrix

File: vulcanai2/models/metrics.py [hunk 4]
```diff
@@ -1,32 +1,66 @@
...
+        return correct, accuracy
 
-    def get_precision(preds, targets):
-        return scipy_metrics.precision_score(targets.flatten(), preds.flatten())
+    def get_precision(predictions, targets):
+        return scipy_metrics.precision_score(targets.flatten(), predictions.flatten())
 
-    def get_roc_score(preds, targets):
-        return scipy_metrics.roc_auc_score(targets.flatten(), preds.flatten())
+    def get_roc_score(predictions, targets):
```
Comment (rfratila) [inline vulcanai2/models/metrics.py]: I think the sklearn roc and abc calculation work better than scipy

File: vulcanai2/models/metrics.py (hunk 3 above)
Comment (nirtiac) [reply vulcanai2/models/metrics.py]: @priyatharsan I think this is relevant to you

File: vulcanai2/datasets/tabulardataset.py [hunk 5]
```diff
@@ -33,48 +42,111 @@ def __getitem__(self, idx):
...
+        """
+        return list(self.df)
 
     def get_all_numeric_features(self):
-        return [key for key in dict(self.data.dtypes) if dict(self.data.dtypes)[key] in ['float64', 'int64']]
+        """
+        Returns all columns that contain numeric values
+        :return: all columns that contain numeric values
+        """
+        return [key for key in dict(self.df.dtypes) if dict(self.df.dtypes)[key] in ['float64', 'int64']]
```
[the remark this answers is not in the document]
Comment (nirtiac) [reply vulcanai2/datasets/tabulardataset.py]: oh, yes!

File: vulcanai2/datasets/utils.py (hunk 2 above)
Comment (nirtiac) [reply vulcanai2/datasets/utils.py]: I can't find that, so I'll just ask Joseph to add it in at the end.

File: vulcanai2/models/metrics.py (hunk 1 above)
Comment (priyatharsan) [reply vulcanai2/models/metrics.py]: added that to the new PR

File: vulcanai2/models/metrics.py (hunk 3 above)
Comment (priyatharsan) [reply vulcanai2/models/metrics.py]: yup fixed it

File: vulcanai2/models/metrics.py (hunk 4 above)
Comment (priyatharsan) [reply vulcanai2/models/metrics.py]: they are actually sklearn methods. I named them wrong, apologies..

**Pull request merged**